File: C++/ex6/longest_increasing_subsequence_solution.cpp
// Solution for the LIS exercise (O(n log n))
// Uses the "tails" technique: for each possible length, keep the smallest
// ending value of an increasing subsequence of that length.

#include <bits/stdc++.h>
using namespace std;

int lis_length(const vector<int>& a) {
    if (a.empty()) return 0;
    vector<int> tails; // tails[len-1] = minimum ending value of an increasing subsequence of length len
    for (int x : a) {
        auto it = lower_bound(tails.begin(), tails.end(), x);
        if (it == tails.end()) tails.push_back(x);
        else *it = x;
    }
    return (int)tails.size();
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n)) return 0;
    vector<int> a(n);
    for (int i = 0; i < n; ++i) cin >> a[i];
    cout << lis_length(a) << '\n';
    return 0;
}

File: C#/ex6/longest_increasing_subsequence_solution.cs
// Solution Ex 6 — Longest Increasing Subsequence (O(n log n))
// Uses tails: tails[len-1] = minimum possible tail value for an increasing
// subsequence of length len.

using System;
using System.Collections.Generic;

class Program {
    static int LisLength(List<int> a) {
        if (a == null || a.Count == 0) return 0;
        var tails = new List<int>();
        foreach (var x in a) {
            int i = tails.BinarySearch(x);
            if (i < 0) i = ~i; // insertion point
            if (i == tails.Count) tails.Add(x);
            else tails[i] = x;
        }
        return tails.Count;
    }

    static void Main() {
        if (!int.TryParse(Console.ReadLine(), out int n)) return;
        var tokens = new List<int>();
        while (tokens.Count < n) {
            string? line = Console.ReadLine();
            if (line == null) break;
            foreach (var t in line.Split(new[]{' ', '\t'}, StringSplitOptions.RemoveEmptyEntries)) {
                if (tokens.Count >= n) break;
                if (int.TryParse(t, out int v)) tokens.Add(v);
            }
        }
        Console.WriteLine(LisLength(tokens));
    }
}
